### PR TITLE
[IMP] server.py: implements limits (real) for multi-threading

### PR DESCRIPTION
**Pull request by beledouxdenis**
[IMP] server.py: implements limits (real) for multi-threading

The multi-threaded server graceful stop policy is now as follows:
 - shutdown the server so it does not accept new requests (the socket remains open),
 - wait for remaining requests to finish up to a certain time (e.g. 5 seconds)
   - requests that already reached a limit are excluded.

Previously, according to the werkzeug version installed, the multi-threaded server waited either indefinitely for requests to finish or not at all.

The reload policy when a thread reaches a limit is as following:
 - wait there is no other running requests, up to 1 minute
 - then stop the server (with the above graceful stop policy applied),
 - then reload the server, therefore accepting new requests again.
- The socket is not closed during this operation.
  This allows the inverse proxy to queue the requests arriving during
  the time the server is shutted down and therefore not accepting new requests.
  This allows a transparent restart for the users, as their
  requests done during the shutting down of the server are not rejected but queued.
- When a thread reaches a limit, we reload the whole threaded server instead of
  just killing this thread. Killing a thread is not a good practice as it could
  lead to dead locks in case these threads acquire locks.
  Killing a thread that acquired a lock would lead for this lock to never be
  released.

Regarding the override of `_handle_request_noblock`:
  In `socketserver`,
  when `shutdown` is called
  `__shutdown_request` is set to True,
  https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L241-L249
  This tells the server to no longer accept requests. Callstack below:
  - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L231-L234
  - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L308
  - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L487

  Unfortunately, this is very possible the HTTP server is already polling for a request when this flag is set
  to `True` by the main thread:
  - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L232
  As the poll_interval is set to 0.5s:
  - https://github.com/python/cpython/blob/6a7b3a77b4b2be0badd24ee5f0fdbaa2e0e79c3d/Lib/socketserver.py#L215
  if a request arrives within 500ms after the server is shut down,
  the request is accepted, while the server is shutdown.

  We want to prevent this: Once the server is marked as shut down, we want it to no longer accept requests:
  so the inverse proxy knows it has to queue the request.

### CODE REVIEW

File: odoo/service/server.py [hunk 1]
```diff
@@ -117,6 +125,55 @@ def server_activate(self):
...
+        """
+        t = threading.Thread(target = self.process_request_thread,
+                             args = (request, client_address))
+        t.daemon = self.daemon_threads
+        t.start()
+        self._request_threads[t] = {'start_time': time.time()}
+
+    def _handle_request_noblock(self):
+        """
+        In `socketserver`,
```
Comment (antonylesuisse) [inline odoo/service/server.py]: What about this more simple text:
in the python socketserver process_request loop the __shutdown_request flag is not check between select and accept  thus when we set it to true a last request is accepted before exiting the loop. We override this function to add an additionalcheck before the accept()

File: odoo/service/server.py [hunk 2]
```diff
@@ -117,6 +125,55 @@ def server_activate(self):
...
+        """
+        if self._BaseServer__shutdown_request:
+            return
+        super(ThreadedWSGIServerReloadable, self)._handle_request_noblock()
+
+    @property
+    def request_threads(self):
+        """
+        A property returning the alive request threads.
+        """
```
Comment (antonylesuisse) [inline odoo/service/server.py]: Why a proprety ?

File: odoo/service/server.py [hunk 3]
```diff
@@ -212,6 +272,51 @@ def signal_handler(self, sig, frame):
             # interrupt run() to start shutdown
             raise KeyboardInterrupt()
 
+    def process_limit(self):
+        rss, vms = memory_info(psutil.Process(os.getpid()))
+        if vms > config['limit_memory_soft']:
+            _logger.info('Thread virtual memory limit (%s) reached.', vms)
+            self.limits_reached_threads.add(threading.currentThread())
+
+        for thread, thread_info in self.threads.items():
```
Comment (antonylesuisse) [inline odoo/service/server.py]: That's O(n**2)

File: odoo/service/server.py (hunk 2 above)
Comment (beledouxdenis) [reply odoo/service/server.py]: I can do a method if you prefer, it was just to make it more elegant when I had to use loop on the threads, but it does make the code a bit more complicated.

File: odoo/service/server.py (hunk 3 above)
Comment (beledouxdenis) [reply odoo/service/server.py]: wut ?

Comment (antonylesuisse) [reply odoo/service/server.py]: Because you use a property you hide the fact that's it's every access to the property is O(n)

Comment (antonylesuisse) [reply odoo/service/server.py]: Let's talk about it IRL.

**Comment by beledouxdenis** [comment]
@antonylesuisse PR updated according to your review : )

**Comment by beledouxdenis** [comment]
@robodoo retry

**Comment by beledouxdenis** [comment]
@robodoo r+

**Pull request closed without merging**